garage/tf/algos/reps.py

```diff
@@ -0,0 +1,442 @@
+        dual_optimzier: dual func optimizer.
+        dual_optimizer_args: arguments of the dual optimizer.
+
+    """
+
+    def __init__(self,
+                 epsilon=0.5,
+                 l2_reg_dual=0.,
+                 l2_reg_loss=0.,
+                 optimizer=LbfgsOptimizer,
+                 optimizer_args=dict(max_opt_itr=50),
+                 dual_optimizer=scipy.optimize.fmin_l_bfgs_b,
+                 dual_optimizer_args=dict(maxiter=50),
+                 name="REPS",
+                 **kwargs):
+        self.name = name
+        self._name_scope = tf.name_scope(self.name)
+
+        with self._name_scope:
+            self.optimizer = optimizer(**optimizer_args)
+            self.dual_optimizer = dual_optimizer
+            self.dual_optimizer_args = dual_optimizer_args
+            self.epsilon = float(epsilon)
+            self.l2_reg_dual = float(l2_reg_dual)
+            self.l2_reg_loss = float(l2_reg_loss)
+
+        super().__init__(**kwargs)
+
+    @overrides
+    def init_opt(self):
+        """Initialize the optimization procedure."""
+        pol_loss_inputs, pol_opt_inputs, dual_opt_inputs = self._build_inputs()
+        self._policy_opt_inputs = pol_opt_inputs
+        self._dual_opt_inputs = dual_opt_inputs
+
+        pol_loss = self._build_policy_loss(pol_loss_inputs)
+        self.optimizer.update_opt(
+            loss=pol_loss,
+            target=self.policy,
+            inputs=flatten_inputs(self._policy_opt_inputs))
+
+    @overrides
+    def get_itr_snapshot(self, itr, samples_data):
+        """Return the data should saved in the snapshot."""
+        return dict(
+            itr=itr,
+            policy=self.policy,
+            env=self.env,
+        )
+
+    @overrides
+    def optimize_policy(self, itr, samples_data):
+        """Perform the policy optimization."""
+        # Initial BFGS parameter values.
+        x0 = np.hstack([self.param_eta, self.param_v])
+        # Set parameter boundaries: \eta>=1e-12, v unrestricted.
+        bounds = [(-np.inf, np.inf) for _ in x0]
+        bounds[0] = (1e-12, np.inf)
+
+        # Optimize dual
+        eta_before = self.param_eta
+        logger.log("Computing dual before")
+        dual_opt_input_values = self._dual_opt_input_values(samples_data)
+        dual_before = self.f_dual(*dual_opt_input_values)
+        logger.log("Optimizing dual")
+
+        def eval_dual(x):
+            self.param_eta = x[0]
+            self.param_v = x[1:]
+            dual_opt_input_values = self._dual_opt_input_values(samples_data)
+            return self.f_dual(*dual_opt_input_values)
+
+        def eval_dual_grad(x):
+            self.param_eta = x[0]
+            self.param_v = x[1:]
+            dual_opt_input_values = self._dual_opt_input_values(samples_data)
+            grad = self.f_dual_grad(*dual_opt_input_values)
+            eta_grad = np.float(grad[0])
+            v_grad = grad[1]
+            return np.hstack([eta_grad, v_grad])
```

Comment:
Should we avoid creating inline functions?

Reply:
We should. But I don't know any simple way to make `eval_dual` and `eval_dual_grad` to be member functions or helper functions outside the class.

Reply:
in this case i think it's fine. it comes down do a design issue with the optimizers